src/operators/mod.rs

```diff
@@ -452,3 +534,255 @@ pub(crate) fn stateful_map(
+        &self,
+        recovery_store_summary: Vec<(StepId, K, S::Timestamp, UpdateType)>,
+        backups: IB,
+        recovery_collector: Rc<Box<dyn RecoveryStore<S::Timestamp, K, D>>>,
+    ) where
+        IB: IntoIterator<Item = Stream<S, (StepId, K, UpdateType)>>;
+}
+
+impl<S, K: ExchangeData + Hash + Eq + Debug> GarbageCollector<S, K> for Stream<S, ()>
+where
+    S: Scope<Timestamp = u64>,
+{
+    fn garbage_collector<IB, D: 'static>(
+        &self,
+        recovery_store_log: Vec<(StepId, K, S::Timestamp, UpdateType)>,
+        backups: IB,
+        recovery_collector: Rc<Box<dyn RecoveryStore<S::Timestamp, K, D>>>,
+    ) where
+        IB: IntoIterator<Item = Stream<S, (StepId, K, UpdateType)>>,
+    {
+        let mut op_builder = OperatorBuilder::new("GarbageCollector".to_string(), self.scope());
+
+        let mut dataflow_frontier_input = op_builder.new_input(self, pact::Pipeline);
+        let mut backup_inputs = backups
+            .into_iter()
+            .map(|s| op_builder.new_input(&s, pact::Pipeline))
+            .collect::<Vec<_>>();
+
+        op_builder.build(move |_init_capabilities| {
+            let mut tmp_recovery_keys = Vec::new();
+            let mut garbage_key_buffer = Vec::new();
+            let mut recovery_store_summary: HashMap<
+                (StepId, K),
+                BTreeSet<(S::Timestamp, UpdateType)>,
+            > = HashMap::new();
+
+            // Load the initial summary of the recovery store into
+            // this special query structure in memory.
+            for (step_id, key, epoch, update_type) in recovery_store_log {
+                recovery_store_summary
+                    .entry((step_id, key))
+                    .or_default()
+                    .insert((epoch, update_type));
+            }
+
+            move |input_frontiers| {
+                let mut dataflow_frontier_input_handle =
+                    FrontieredInputHandle::new(&mut dataflow_frontier_input, &input_frontiers[0]);
+
+                // Gotta drain the data on this input to advance its
+                // frontier, even if we do nothing with it.
+                dataflow_frontier_input_handle.for_each(|_cap, _data| {});
+
+                for input in backup_inputs.iter_mut() {
+                    input.for_each(|cap, data| {
+                        data.swap(&mut tmp_recovery_keys);
+                        let epoch = cap.time();
+
+                        // Drain items so we don't have to re-allocate.
+                        for (step_id, key, update_type) in tmp_recovery_keys.drain(..) {
+                            recovery_store_summary
+                                .entry((step_id, key))
+                                .or_default()
+                                .insert((epoch.clone(), update_type));
+                        }
+                    });
+                }
+
+                // TODO: This runs on every operator activation. We
```

Comment:
I'm not totally sure on the call stack that leads to this closure being called. But perhaps I'm being fast and loose r.e. "scheduling" and "activation". My understanding was that it was called on what Frank is calling every scheduling. Our goal is to only run this code on 2, but I assume it also runs on 1 & 3 because that's what a generic operator would need. Hence the desire to have an extra check.

Does that answer your question?